### ISSUE TITLE
Getting entities in just loaded chunk returns an empty collection

### ISSUE THREAD

**Issue by kennytv** [PaperMC/Paper#5872]
So that it's documented before someone forgets:
https://hub.spigotmc.org/jira/browse/SPIGOT-6526 was fixed on Spigot and is confirmed to still be unfixed on Paper, likely due to the async chunk loading patch.

Sync loading a chunk and getting the entities returns an empty collection.
```java
    @Override
    public void onEnable() {
        World world = Bukkit.getWorlds().get(0);
        Chunk chunk = world.getChunkAt(world.getSpawnLocation()); // Sync load
        System.out.println("Chunk loaded: " + chunk.isLoaded() + " entities: " + chunk.getEntities().length); // will be 0, even if entities exist in the chunk

        // Spawn an entity for next startup
        String name = "ID: " + System.currentTimeMillis();
        ArmorStand itemFrame = (ArmorStand) world.spawnEntity(world.getSpawnLocation(), EntityType.ARMOR_STAND);
        itemFrame.setCustomName(name);
        itemFrame.setCustomNameVisible(true);

        // See what happens after a manual unload/load
        System.out.println("Chunk loaded: " + chunk.isLoaded() + " " + chunk.getEntities().length);
        boolean unloaded = chunk.unload();
        System.out.println("Chunk unloaded: " + unloaded);
        chunk.load();
        System.out.println("Chunk loaded: " + chunk.isLoaded() + " " + chunk.getEntities().length);
        System.out.println("Please restart server");
    }
```

**Comment by Foorcee** [comment]
I debugged it further and it does behave a little differently. 
I have disabled keep-spawn-loaded and deleted the world.

**First run:**

```
[11:17:14 INFO]: Enabling TestEntities v1.0*
[11:17:14 INFO]: Chunk loaded: true entities: 0
[11:17:14 INFO]: Chunk loaded: true 1
[11:17:14 INFO]: Chunk unloaded: false
[11:17:14 INFO]: Chunk loaded: true 1
[11:17:14 INFO]: Please restart server
```

New world was generated. Everything looks good so far. At the beginning there are no entities.After spawning the armor stand, one entity is present.
I don't know why the chunk can't be unloaded.

**Second run:**

```
[11:22:55 INFO]: Enabling TestEntities v1.0*
[11:22:56 INFO]: Chunk loaded: true entities: 0
[11:22:56 INFO]: Chunk loaded: true 0
[11:22:56 INFO]: Chunk unloaded: false
[11:22:56 INFO]: Chunk loaded: true 0
[11:22:56 INFO]: Please restart server
```

World is loaded. Again no entities are found and also after spawning another armor stand, the entities list is empty.

Both armor stands are present when I join the server:

![img](https://i.imgur.com/nQ8sNpO.png)

My Paper config:
http://paste.cytooxien.de/gahiqiwimu.coffeescript

Tested version:

This server is running Paper version git-Paper-"ea0336d" (MC: 1.17) (Implementing API version 1.17-R0.1-SNAPSHOT)

**Comment by electronicboy** [comment]
This behavior is 100% intentional by vanilla, that's the entire issue here

**Comment by Foorcee** [comment]
This might be possible, it doesn't bother in Vanilla, but if you want to remove all entities from a chunk it is rather impractical. Besides, it was fixed in Spigot. Also, I can hardly imagine that it is an intended behavior that after spawning an entity, the list is still empty. 
So basically, I think the feature is good. But when I load a chunk sync or it is already loaded, I expect that it is also fully loaded, which means that also the entities from the chunk are loaded at that time.

**Comment by electronicboy** [comment]
To my understanding, it wasn't really fixed in spigot is the issue, they fixed for spawn loading but that's purely relying on loading chunks to be finished before after of the entities are actually loaded ready to be added to the world, they've not actually resolved the underlying issue is the problem

Blocking to load entities for sync loads somewhat makes sense from an API perspective but is not really suitable as a full term solution across the board

**Comment by Brokkonaut** [comment]
Would something like Chunk.loadEntities() and/or Chunk.loadEntitiesAsync() to force load entities in that chunk and maybe some ChunkLoadEntitiesEvent/ChunkUnloadEntitiesEvent as a replacement possible?

**Comment by electronicboy** [comment]
I mean, I question if getChunk (or more specifically getEntities in the API) in the API should attempt to forcefully load entities, but that's a pretty shitty solution too, separate events would be ideal in allowing the existing vanilla flow to continue

**Comment by Jeppa** [comment]
> ...Besides, it was fixed in Spigot. Also, I can hardly imagine that it is an intended behavior that after spawning an entity, the list is still empty.

see also here: https://hub.spigotmc.org/jira/browse/SPIGOT-6547

**Comment by molor** [comment]
Is there any news or new events from which we can get the loaded entities?..

**Comment by Machine-Maker** [comment]
Spigot has an [open PR](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/pull-requests/866/overview) for this, I'm not sure its a good idea for Paper to add stuff if spigot is going to add their equivalents right after.

**Comment by molor** [comment]
> [open PR](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/pull-requests/866/overview)

I don't have permission to access it, but OK, thanks

**Comment by Brokkonaut** [comment]
> > [open PR](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/pull-requests/866/overview)
> 
> I don't have permission to access it, but OK, thanks

you would have to sign the spigot cla and be logged in to have access to it. it is a bit strange that one cannot even view PRs without the cla there...

**Comment by Novalescent** [comment]
Is there any known workaround to this? I tried doing /forceload on chunks and not even those returned the mobs.

**Comment by Diranus** [comment]
Unfortunately, I only found this clumsy workaround to circumvent this so far: Reloading the affected plugins after first join using a plugin manager. I don't want to do this in our production system.

So what is the status of this bug? It is really annoying and in my case, it blocks a bigger project from being updated.

Is there a more reliable workaround? 

Other Bukkit implementations (Spigot, CraftBukkit) don't have this issue at all.

**Comment by electronicboy** [comment]
This behaviour is from spigot...

Current workaround is to use the entity add to world events

**Comment by Diranus** [comment]
> This behaviour is from spigot...

No, it doesn't seem to be the case anymore - Spigot just has fixed this bug recently. ;)

I was more or less forced to test Spigot's behavior by temporarely switching back my development system from Paper to regular Spigot a few days ago in order to continue developing on the 1.17 update of my plugins while waiting for this issue gets fixed in Paper too. 

After I did that, I didn't encounter any issue like this at all. The collection mentioned above is not empty. It seems to contain all the entities loaded before the onEnable() of my plugin is reached, just like it was in 1.16.

In my opinion, this bug seriously breaks the backward compatibility to Spigot- and Bukkit-plugins. :-(

Please, please fix it....

**Comment by electronicboy** [comment]
Spigot hijacks this to work for onEnable, it does not work past the world loading, "fixing" this would require making chunk loading sync again on startup, or fixing the core underlying issue which causes this (Which is basically an upstream issue and they have a PR in the pipeline for it but don't seem to be ultra interested in merging it yet), especially when alternatives exist which don't impact the server startup

**Comment by Jeppa** [comment]
> No, it doesn't seem to be the case anymore - Spigot just has fixed this bug recently. ;)

No, it hasn't...

> This behaviour is from spigot... Current workaround is to use the entity add to world events

This might work, if you need the entity itself...
But, if you need the amount of entities in a chunk/world that won't help...

**Comment by Combustible** [comment]
TLDR: +1

Getting entities after a chunk is loaded (both sync and async) is also very important for us and blocks upgrading to 1.17. 

We have two different categories of systems that sound like they will be broken by this - I haven't tested yet. 

- We catch ChunkLoadEvent to put various entities into various trackers. This is is used in a lot of places, especially for stuff like armor stands with special properties, tracking boats/minecarts, etc. 
- We have one block of code that loads chunks async then immediately iterates and removes(!) entities that match particular criteria. This is part of a routine that copy/pastes over a section of the world - we first load the chunks, then remove entities, then paste new blocks and entities on top of that area, of the form: 

```java
Consumer<Chunk> chunkConsumer = (Chunk chunk) -> {
        for (BlockState state : chunk.getTileEntities(true)) {
                ...
        }

        for (Entity entity : chunk.getEntities()) {
                if (box.contains(entity.getLocation().toVector()) && entityShouldBeRemoved(entity)) {
                        ...
                }
        }
};

/* Load all the chunks in the region and run the chunk consumer */
for (BlockVector2 chunkCoords : chunks) {
        world.getChunkAtAsync(chunkCoords.getX(), chunkCoords.getZ(), chunkConsumer);
}```

**Comment by Machine-Maker** [comment]
Upstream supposedly addressed this [here](https://hub.spigotmc.org/stash/projects/SPIGOT/repos/craftbukkit/commits/15f27fc710eb9207983147f48f4d51602c34749b) but we've got an open issue regarding that logic just added, so I'll leave this open until more people chime in with results

**Comment by TheFlash787** [comment]
Has there been an update to this? I'm getting the same issue when using `getNearbyEntities` on a location - either with a radius or a bounding box.

**Comment by electronicboy** [comment]
I mean, are you calling getNearbyEntities on a freshly loaded chunk? if so, it's this, otherwise, it's not

upstream looks to have maybe addressed the reason why I reverted their fix, needs looking into during the next upstream

**Comment by TheFlash787** [comment]
> 
> 
> I mean, are you calling getNearbyEntities on a freshly loaded chunk? if so, it's this, otherwise, it's not
> 
> upstream looks to have maybe addressed the reason why I reverted their fix, needs looking into during the next upstream

Not grabbing directly from the freshly loaded chunk, no. However I can create a new issue to diagnose my specific case further if this is not directly related (although may be indirectly)

**Comment by electronicboy** [comment]
if it's not a chunk which has basically just loaded then it's nothing to do with this issue

**Comment by Machine-Maker** [comment]
https://github.com/Machine-Maker/Paper/blob/fix/CraftChunk%23getEntities/patches/server/0838-improve-CraftChunk-getEntities.patch is a solution I came up with, it passed basic testing.

**Comment by Jeppa** [comment]
Will this ever be fixed?
In Spigot itself there is a fix for this, ok it's not 100%, but way better than no fix at all...

**Comment by electronicboy** [comment]
Yes, one day, over the rainbow. Spigots fix was removed because it was crashing the server.

**Comment by Machine-Maker** [comment]
If someone wants to test https://github.com/Machine-Maker/Paper/tree/fix/CraftChunk%23getEntities this, its my potential fix for it, idk how good it is, like I posted above, it passed basic testing, but there are prob still some issues.

**Comment by Jeppa** [comment]
in 1.19 it's still the same... :(

**Comment by Beafty** [comment]
did it get fix in 1.18.2 now? slimfun4 hologram almost made me mad :<

**Comment by electronicboy** [comment]
This is intended behavior from mojang, there is an in progress thing to attempt to mitigate this, but, there's already better options for plugin devs to use here

**Comment by Jeppa** [comment]
> 

A better option for this ? What can be better than chunk#getEntities() or world#getEntities() ?
And please don't talk about events, as they can't be used if I need a sub with a return value...

**Comment by electronicboy** [comment]
getEntities on a chunk without the entities loaded requires a blocking load, the reason we pulled spigots hackfix out of paper is because it kept killing servers running paper due chunks loading faster; any "fix" to this is basically hiding how mojangs system now works and is practiced in some sense in the long run, a giant potential headache. You'd be much better off actually adapting to how the modern stack works in the long run rather than trying to run away from it.

**Comment by Brokkonaut** [comment]
What I am still missing in the API is just some World.loadChunkEntites() and/or better World.loadChunkEntitesAysnc() method, to be able to force load entities in a chunk when i want to manipulate them.

**Comment by Felina-Lain** [comment]
Ack, that explains the hours I spent trying to make my plugin work.
Any workaround has been found? Or do I need to switch to spigot?

**Comment by electronicboy** [comment]
This has been fixed in the recent builds to my understanding

**Comment by lynxplay** [comment]
I just ran the example code provided, cats understanding is correct. This has been fixed with leafs chunk rewrite.
I am closing this issue.
